### PR TITLE
Ensure that users can only create workers with scopes that they have

### PR DESCRIPTION
**Pull request by jhford**
Including unit tests!

### CODE REVIEW

**Comment by jhford** [comment]
@jonasfj can you do a pretty good check on the scopes method that I added, i want to be extra certain that I didn't goof on the logic!

File: lib/worker-type.js [hunk 1]
```diff
@@ -339,6 +339,59 @@ WorkerType.prototype.testLaunchSpecs = function () {
...
+ * nice if we could convert the array of arrays into a set of sets for
+ * comparisons
+ */
+WorkerType.scopes = function (worker) {
+  // Generate the complete list of scopes;
+  let scopes = [];
+
+  // Take an object with a .scopes property and list add all of its (mostly)
+  // unique scopes to the list of scopes to return
+  var y = function(x) {
```
Comment (jhford) [inline lib/worker-type.js]: ugh, that should be a let not a var

File: lib/routes/v1.js [hunk 2]
```diff
@@ -105,6 +105,15 @@ api.declare({
     return;
   }
 
+  // Figure out which scopes the client needs
+  let requiredScopes = this.WorkerType.scopes(input);
+  // Now check for them if there are any
+  if (requiredScopes.length > 0) {
+    if (!req.satisfies(requiredScopes)) {
```
Comment (jonasfj) [inline lib/routes/v1.js]: I think this has to be `req.satisfies([requiredScopes])`..

The argument given to `satisfies` is a set of scope-sets in disjunctive form.
e.g. `req.satisfies([[a], [b,c]])` means "`a` or (`b` and `c`)"

File: lib/worker-type.js [hunk 3]
```diff
@@ -339,6 +339,59 @@ WorkerType.prototype.testLaunchSpecs = function () {
...
+    y(instanceType);
+  }
+
+  return scopes; 
+};
+
+/**
+ * Make the scopes method available on instances as well
+ */
+WorkerType.prototype.scopes = function () {
```
Comment (jonasfj) [inline lib/worker-type.js]: This overwrites a property... So probably a bad idea...

File: lib/routes/v1.js (hunk 2 above)
Comment (jhford) [inline lib/routes/v1.js]: if I'm remembering the code I wrote properly, `requiredScopes` will have that form.

From the later test:
```javascript
+      subject.scopes(wType).should.eql([
+        ['gScope', 'g'],
+        ['rScope1', 'r'],
+        ['rScope2'],
+        ['r'],
+        ['tScope1', 't'],
+        ['tScope2'],
+        ['t'],
+      ]);
```

File: lib/worker-type.js (hunk 3 above)
Comment (jhford) [inline lib/worker-type.js]: omg.  yes.  oops.

File: lib/worker-type.js [hunk 4]
```diff
@@ -339,6 +339,59 @@ WorkerType.prototype.testLaunchSpecs = function () {
...
+    y(region);
+  }
+
+  // Instance Type specific scopes
+  for (let instanceType of worker.instanceTypes) {
+    y(instanceType);
+  }
+
+  return scopes; 
+};
```
Comment (jonasfj) [inline lib/worker-type.js]: This seems way to complicated... and wrong.
Just do:
```js
return _.union(
  worker.scopes,
  _.flatten(worker.instanceTypes.map(t => t.scopes)),
  _.flatten(worker.regions.map(r => r.scopes))
]));
```
I assume you're code always creates the `scopes` property as an array, even if it's empty.
If you don't do that... then I recommend that you start doing so, and do a version upgrade then you can fix the old instances in the `migrate` function.

Sure you could add `|| []` everywhere but that is bad... Defaults should exists in JSON schemas and if not there, then when objects are created. But I'm pretty sure you can settle with just having them in your JSON schemas (for the input).

**Comment by jonasfj** [comment]
@jhford, has some variant of this been merged?
If not the comments seems easy...

**Pull request closed without merging**